### PR TITLE
IF-ELSE support for conditional nodes.

### PR DESCRIPTION
**Pull request by galv**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #189462
* #189461
* #186056

Requires CUDA 12.8 or higher. Slightly faster than using two IF nodes.

### CODE REVIEW

File: torch/_higher_order_ops/cudagraph_conditional_nodes.py [hunk 1]
```diff
@@ -5,9 +5,15 @@
 import torch
 import torch.utils._pytree as pytree
 from torch._higher_order_ops.auto_functionalize import auto_functionalized_v2_dense
+from torch.torch_version import TorchVersion
 from torch.utils._python_dispatch import TorchDispatchMode
 
 
+_CUDA_GRAPH_IF_ELSE_NODES_SUPPORTED = (
```
Comment (aorenste) [inline torch/_higher_order_ops/cudagraph_conditional_nodes.py:12]: [blocker] My agent says:
> Gate on a capability query that ANDs the compile-time `CUDA_VERSION >= 12080` guard with `cudaDriverGetVersion() >= 12080`, and fall back to the two-IF path when it fails.
>
> `torch.version.cuda` is the toolkit PyTorch was *built* with, but `conditional.size = 2` is validated by the driver. Under CUDA minor-version compatibility a 12.8/12.9-toolkit build can run on a pre-r570 driver, where `cudaGraphAddNode` rejects the two-body IF and capture hard-errors -- a configuration that worked before via the two-IF fallback. (CUDA 13.x needs r580+, so it is not affected.) `torch/cuda/green_contexts.py:61` already does the driver half of this check for the same 12.8 feature class.
>
> Secondary benefit: this predicate currently duplicates the `CUDA_VERSION >= 12080` guard in `begin_capture_to_if_else_node`, so the two can drift.
>
> I did not test this on a pre-r570 driver.

File: torch/_higher_order_ops/cudagraph_conditional_nodes.py [hunk 2]
```diff
@@ -229,13 +235,39 @@ def _if_body(pred: torch.Tensor) -> Generator[None, None, None]:
...
+    current_cuda_graph = torch.cuda.CUDAGraph.get_currently_capturing_graph()
+    current_cuda_graph.begin_capture_to_if_else_node(pred)
+    try:
+        yield current_cuda_graph
+    finally:
+        current_cuda_graph.end_capture_to_conditional_node()
+
+
+def _copy_else_branch_outputs(if_out, else_out) -> None:
+    for if_out_leaf, else_out_leaf in zip(
```
Comment (aorenste) [inline torch/_higher_order_ops/cudagraph_conditional_nodes.py:249]: [cleanup] My agent says:
> Same `copy_` broadcast hole as `switch_node` in #189461: branches differing in a size-1 dim silently fill the if-branch buffer instead of failing -- `true_fn` returning `(4,)` and `false_fn` returning `(1,)` gives four copies of the single value at replay. Wants a shape check here, plus `strict=True` on the `zip` for the leaf-count case.
>
> Nothing upstream rejects it: `_merge_output` (`cond.py:452`) checks `dim` but explicitly not size -- same-rank divergence merges to an unbacked SymInt -- and the dispatch mode intercepts `torch.ops.higher_order.cond` before any of that runs. A captured graph has fixed sizes and cannot represent a branch-dependent output shape.

File: test/functorch/test_control_flow.py [hunk 3]
```diff
@@ -7293,6 +7294,64 @@ def f(x, y):
...
+        graph = torch.cuda.CUDAGraph(keep_graph=True)
+        with (
+            torch.cuda.graph(graph, stream=side_stream),
+            CUDAGraphCaptureControlFlowOpDispatchMode(),
+        ):
+            out = f(pred, x)
+
+        torch.cuda.current_stream().wait_stream(side_stream)
+
+        if TEST_CUDA_PYTHON_BINDINGS:
```
Comment (aorenste) [inline test/functorch/test_control_flow.py:7327]: [cleanup] My agent says:
> Replace this runtime guard with `@requires_cuda_python_bindings` (`torch/testing/_internal/common_utils.py:2606`) on the test, so a missing `cuda.bindings` is reported as a skip.
>
> As written, the node-count assertion -- the only thing separating this from the two-IF path -- is silently dropped without the bindings and the test still reports pass. Plain cond-under-cudagraph replay correctness is already covered by `test_cond_traced_not_nested_cudagraphs`.

**Review by aorenste — CHANGES_REQUESTED** [review]
(Reviewed by me, assisted by AI)

[blocker] My agent says:
> Nothing in CI exercises the pre-12.8 two-IF fallback any more. Every CUDA test job in `pull.yml` / `trunk.yml` / `periodic.yml` / `inductor*.yml` / `slow.yml` builds against CUDA 13.0-13.4, so from this change on `if_else_node` always takes the new single-node path; before it, the fallback was the only path and every `test_cond_*_cudagraphs` test covered it. It still ships: `cmake/public/cuda.cmake:73` accepts CUDA 12.6+, and cu126 is one of the four builds in `generate_binary_build_matrix.py:27`.
>
> In the tests, add a mock overriding `_CUDA_GRAPH_IF_ELSE_NODES_SUPPORTED` to cover the pre-12.8 path. It needs to wrap the warmup as well as the capture, since `ControlFlowOpWarmupDispatchMode` also calls `if_else_node`.
